### PR TITLE
Add full support for ragg

### LINKED ISSUE #4347: ragg device functions not working with ggsave
In the recent tidyverse blogpost "[Modern Text Features in R](https://www.tidyverse.org/blog/2021/02/modern-text-features/)" it states you can use the ragg device functions with ggsave:

> You can use ragg with ggsave() by passing the device function to the device argument (e.g.  ggsave(device = agg_tiff)).

It's not working for me however, the following code results in an empty image being saved.

Related to https://github.com/r-lib/ragg/issues/69

``` r
library(ggplot2)
library(ragg)

p <- ggplot(iris, aes(Sepal.Length, Sepal.Width, col = Species)) + geom_point() 

ggsave("gg_test.png", p, device = agg_png) 
#> Saving 7 x 5 in image
```

<sup>Created on 2021-02-17 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.3 (2020-10-10)
#>  os       macOS Mojave 10.14.6        
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_GB.UTF-8                 
#>  ctype    en_GB.UTF-8                 
#>  tz       Europe/Paris                
#>  date     2021-02-17                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source             
#>  assertthat    0.2.1   2019-03-21 [1] standard (@0.2.1)  
#>  backports     1.2.0   2020-11-02 [1] standard (@1.2.0)  
#>  cli           2.2.0   2020-11-20 [1] standard (@2.2.0)  
#>  colorspace    2.0-0   2020-11-11 [1] standard (@2.0-0)  
#>  crayon        1.3.4   2017-09-16 [1] standard (@1.3.4)  
#>  digest        0.6.27  2020-10-24 [1] standard (@0.6.27) 
#>  dplyr         1.0.2   2020-08-18 [1] standard (@1.0.2)  
#>  ellipsis      0.3.1   2020-05-15 [1] standard (@0.3.1)  
#>  evaluate      0.14    2019-05-28 [1] standard (@0.14)   
#>  fansi         0.4.1   2020-01-08 [1] standard (@0.4.1)  
#>  farver        2.0.3   2020-01-16 [1] standard (@2.0.3)  
#>  fs            1.5.0   2020-07-31 [1] standard (@1.5.0)  
#>  generics      0.1.0   2020-10-31 [1] standard (@0.1.0)  
#>  ggplot2     * 3.3.3   2020-12-30 [1] standard (@3.3.3)  
#>  glue          1.4.2   2020-08-27 [1] standard (@1.4.2)  
#>  gtable        0.3.0   2019-03-25 [1] standard (@0.3.0)  
#>  highr         0.8     2019-03-20 [1] standard (@0.8)    
#>  htmltools     0.5.1.1 2021-01-22 [1] standard (@0.5.1.1)
#>  knitr         1.30    2020-09-22 [1] standard (@1.30)   
#>  labeling      0.4.2   2020-10-20 [1] standard (@0.4.2)  
#>  lifecycle     0.2.0   2020-03-06 [1] standard (@0.2.0)  
#>  magrittr      2.0.1   2020-11-17 [1] standard (@2.0.1)  
#>  munsell       0.5.0   2018-06-12 [1] standard (@0.5.0)  
#>  pillar        1.4.7   2020-11-20 [1] standard (@1.4.7)  
#>  pkgconfig     2.0.3   2019-09-22 [1] standard (@2.0.3)  
#>  purrr         0.3.4   2020-04-17 [1] standard (@0.3.4)  
#>  R6            2.5.0   2020-10-28 [1] standard (@2.5.0)  
#>  ragg        * 1.1.0   2021-02-15 [1] standard (@1.1.0)  
#>  reprex        1.0.0   2021-01-27 [1] standard (@1.0.0)  
#>  rlang         0.4.10  2020-12-30 [1] standard (@0.4.10) 
#>  rmarkdown     2.5     2020-10-21 [1] standard (@2.5)    
#>  rstudioapi    0.13    2020-11-12 [1] standard (@0.13)   
#>  scales        1.1.1   2020-05-11 [1] standard (@1.1.1)  
#>  sessioninfo   1.1.1   2018-11-05 [1] standard (@1.1.1)  
#>  stringi       1.5.3   2020-09-09 [1] standard (@1.5.3)  
#>  stringr       1.4.0   2019-02-10 [1] standard (@1.4.0)  
#>  styler        1.3.2   2020-02-23 [1] standard (@1.3.2)  
#>  systemfonts   1.0.1   2021-02-09 [1] standard (@1.0.1)  
#>  textshaping   0.3.0   2021-02-10 [1] standard (@0.3.0)  
#>  tibble        3.0.4   2020-10-12 [1] standard (@3.0.4)  
#>  tidyselect    1.1.0   2020-05-11 [1] standard (@1.1.0)  
#>  vctrs         0.3.5   2020-11-17 [1] standard (@0.3.5)  
#>  withr         2.3.0   2020-09-22 [1] standard (@2.3.0)  
#>  xfun          0.19    2020-10-30 [1] standard (@0.19)   
#>  yaml          2.2.1   2020-02-01 [1] standard (@2.2.1)  
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

Thanks.

### PR DESCRIPTION
**Pull request by thomasp85**
Fix #4347 

This PR does two things: First, it will use ragg as the default device if it is installed when generating png, jpeg, and tiff files. Second, it fixes the issue that has crippled the use of `device = agg_png` in ggsave, in that it explicitly sets the resolution and units to use in the device function that is passed in. While this only works for devices that uses `res` and `units` arguments it means that devices that uses the same signature as `png()` will work the same way

### CODE REVIEW

File: R/save.r [hunk 1]
```diff
@@ -158,18 +158,34 @@ plot_dev <- function(device, filename = NULL, dpi = 300) {
...
+    args <- formals(device)
+    call_args <- list()
+    if ("file" %in% names(args)) {
+      call_args$file <- filename
+    }
+    if ("res" %in% names(args)) {
+      call_args$res <- dpi
+    }
+    if ("units" %in% names(args)) {
+      call_args$units <- 'in'
```
Comment (clauswilke) [inline R/save.r:170]: Is this correct? Shouldn't the `units` argument be handed through if the user provides a generic device function?

**Comment by clauswilke** [comment]
While you're at it, I would suggest to add at least the unit `px`, and maybe `pt` also. It seems strange to me that we can't specify output in `px`, the most commonly used unit in web design.

File: R/save.r (hunk 1 above)
Comment (thomasp85) [reply R/save.r:170]: Yes, this is correct. `ggsave()` converts the input and will always serve the dimensions as inches to the graphics device. This was the root cause of the whole ragg debacle

**Comment by thomasp85** [comment]
Yeah — px should at least be supported. I don't think pt has much merit as an output dimension unit?

**Comment by clauswilke** [comment]
Normally my attitude would be "it doesn't cost much to add pt so why not" but I can see that people might get confused about the difference between px and pt so maybe better leave it out.

**Comment by thomasp85** [comment]
Yeah, that is kinda my worry as well

File: R/save.r (hunk 1 above)
Comment (clauswilke) [reply R/save.r:170]: Ah, I see.

**Comment by thomasp85** [comment]
I've added px support - @clauswilke can you give this a review?

**Review by clauswilke — APPROVED** [review]

**Pull request merged**